notebooks/value_semantics/01_value_semantics_in_swift.ipynb

```diff
@@ -0,0 +1,949 @@
+    "\n",
+    "They work exactly as you'd expect."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "1\r\n",
+      "2\r\n"
+     ]
+    }
+   ],
+   "source": [
+    "func incrementTwoInts(_ a: inout Int, _ b: inout Int) {\n",
+    "    a += 1\n",
+    "    b += 1\n",
+    "}\n",
+    "\n",
+    "var a = 0\n",
+    "var b = 1\n",
+    "incrementTwoInts(&a, &b)\n",
+    "print(a)\n",
+    "print(b)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Mutating Functions\n",
+    "\n",
+    "Object-oriented programming taught us that we often want to group data and operations together. For example, we often like to write:\n",
+    "\n",
+    "```python\n",
+    "myType.method()\n",
+    "```\n",
+    "\n",
+    "instead of\n",
+    "\n",
+    "```python\n",
+    "top_level_function(myType)\n",
+    "```\n",
+    "\n",
+    "The extension for `inout` to \"methods\" is spelled `mutating`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "struct Foo {\n",
+    "    var x: Int = 0\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "extension Foo {\n",
+    "    mutating func incrementByOne() {\n",
+    "        x += 1\n",
+    "    }\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Foo(x: 0)\r\n",
+      "Foo(x: 1)\r\n"
+     ]
+    }
+   ],
+   "source": [
+    "func testFunction() {\n",
+    "    var f = Foo()\n",
+    "    print(f)\n",
+    "    f.incrementByOne()\n",
+    "    print(f)\n",
+    "}\n",
+    "testFunction()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Naturally, mutating functions can't be called on immutable values."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [
+    {
+     "ename": "",
+     "evalue": "",
+     "output_type": "error",
+     "traceback": [
+      "error: <Cell 11>:4:7: error: cannot use mutating member on immutable value: 'f' is a 'let' constant\n    f.incrementByOne()  // Compile error\n    ~ ^\n\n<Cell 11>:2:5: note: change 'let' to 'var' to make it mutable\n    let f = Foo()\n    ^~~\n    var\n\n"
+     ]
+    }
+   ],
+   "source": [
+    "func testFunc2() {\n",
+    "    let f = Foo()\n",
+    "    print(f)\n",
+    "    f.incrementByOne()  // Compile error\n",
+    "}\n",
+    "testFunc2()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Mutating functions are convenient in a variety of circumstances. For example, when using an iterator."
```

Comment:
<strike>I have a small concern about using iterator to demonstrate value semantics, since copyability is a key element that makes value semantics interesting, just like mutability.  A noncopyable (always-uniquely-referenced) mutable reference type also has value semantics.  And copyability is ultimately the wrong model for iterator.  But maybe</strike> if they are thinking of the iterator as traversing a multi-pass thing, it's just fine